walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,116 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+
+  // Takes in a collection of all events (each with seperate attendees and timeranges) and
+  // 'MeetingRequest' (with required attendees and a duration requirement) and returns a Collection
+  // of TimeRanges that work with schedules of required attendees and meets the duration requirement
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> result;
+
+    ArrayList<TimeRange> conflicts = generateConflicts(events, request.getAttendees());
+
+    if (conflicts.isEmpty()) {
+      // If no conflicts, create an time range for entire day and add to result if possible (meets
+      // duration requirement)
+      result = new ArrayList<TimeRange>();
+      TimeRange fullDay =
+          TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true);
+      addIfPossible(result, fullDay, request.getDuration());
+    } else {
+      // if conflicts exist, generate free time ranges taking into account conflicts / minimum
+      // duration
+      result = generateFreeTimeRanges(conflicts, request.getDuration());
+    }
+    return result;
+  }
+
+  // Generates arraylist of conflicting events (events attendees of requested meeting are also
+  // attending)
+  private ArrayList<TimeRange> generateConflicts(
+      Collection<Event> events, Collection<String> attendees) {
+    ArrayList<TimeRange> conflicts = new ArrayList<TimeRange>();
+
+    for (Event possibleConflict : events) {
+      // checks if set of attendees of the current 'possible conflicting event' and attendees to
+      // requested event are disjoint (no shared members)
+      if (!Collections.disjoint(possibleConflict.getAttendees(), attendees)) {
+        // adds to conflicts if not disjoint (member in possible conflict is attending requested
```

Comment:
It seems like many comments in this class are self-explanatory based on the code they are describing, including this line. When that's the case, it's best to leave them out entirely. To quote an internal best practices guide about comments:

"What comments are for
Use a comment when it is infeasible to make your code self-explanatory. Here's a great quote from Robert C. Martin's Clean Code:

The proper use of comments is to compensate for our failure to express ourself in code. Note that I used the word failure. I meant it. Comments are always failures. We must have them because we cannot always figure out how to express ourselves without them, but their use is not a cause for celebration.... Every time you express yourself in code, you should pat yourself on the back. Every time you write a comment, you should grimace and feel the failure of your ability of expression.

Self-explanatory code beats liberally-commented code (which, of course, beats code that can't be understood). If you need a comment to explain what a block of code does, try to introduce an explaining variable or to extract a method. If you've already done that but you still need a comment, try to rename it until it answers all your concerns. Sometimes, just choosing better identifier names can be enough; sometimes it takes more work. It's almost always worth it."

I think it's worth going through your code here and removing comments when what they describe is apparent from the code they are attached to.